examples/advanced/nce_example.py

```diff
@@ -0,0 +1,237 @@
+            - c = values
+            - d = d
+        @param x_d: A feasable solution found in the foil set X_v
+        @param foil_idx: A vector containing the items on which the user asked an explanation
+                         All other items must retain their original values
+    """
+    if verbose:
+        print(f"\n\n{'='*10} Solving the master problem {'='*10}")
+    known_solutions, i= [], 1
+    while i:
+        if verbose:
+            print(f"\nStarting iteration {i}")
+        d = intvar(0,INFINITY, values.shape, name="d")
+        x = boolvar(shape=len(x_d), name="x")
+        # The ususal knapsack constraint
+        constraints = [sum(x * weights) <= capacity]
+
+        # Extension of the knapsack problem to the master problem
+        # Ensure the newly found values vector results in a better solution than any known solutions
+        constraints += [sum(d * x) >= sum(d * known) for known in known_solutions]
+        # Ensure values are only modified at foil indices 
+        constraints += [d[i] == values[i] for i in range(len(values)) if i not in foil_idx]
+        # Ensure the foil values assigned by the user remain the same
+        constraints += [x[i] == bool(x_d[i]) for i in foil_idx]
+       
+        master_model = Model(
+            constraints, 
+            # Minimize the change to the values vector
+            minimize = np.linalg.norm(values -  d, ord=1)
+        )
+
+        if master_model.solve() is not False:
+            d_star = d.value()
+            new_solution = solve_knapsack_problem(d_star, weights, capacity)
+            if verbose:
+                print(f"d* = {d_star}")
+                print(f"d* * x_d = {sum(d_star * x_d)}")
+                print(f"d* * x_0 = {sum(d_star * new_solution)}")
+            if sum(d_star * x_d) >= sum(d_star * new_solution):
+                return d_star
+            else:
+                known_solutions.append(new_solution)
+            i += 1
+
+        else:
+            print("Model is UNSAT!")
+            exit()
+
+def solve_knapsack_problem(values, weights, capacity):
+    """
+        Ordinary 0-1 knapsack problem
+        Solve for vector x ∈ {T,F}^n
+
+        Based on the Numberjack model of Hakan Kjellerstrand
+    """
+    x = boolvar(len(values), name="x")
+    model = Model(
+                [sum(x * weights) <= capacity],
+                maximize = sum(x * values)
+            )
+    if model.solve() is not False:
+        return x.value()
+
+
+############################################################################
+# All functions below this line are purely used for pretty printing results#
+############################################################################
+def print_knapsack_model(values, weights, capacity, x):
+    """
+        Pretty prints a knapsack model.
+    """
+    print("Solution to the following knapsack problem")
+    print(f"Capacity: {capacity}, used: {sum(x*weights)}")  
```

Comment:
When running it, I did not know what 'used' was, it is unusual to already use the solution before you present it...

move used to below 'is:' and obj value